### DISCUSSION TITLE
Can we change the feature of the objects in scenes ?

### DISCUSSION THREAD

**Discussion by liuxz-cs** [allenai/ai2thor#535] [Q&A] [answered]
I want to know that if we can change the texture or color or other feature of the objects in scenes. If can, how can we do that ?
And also, I want to know that if we can add some objects into scenes. If can, how can we do that ?

**Comment by mattdeitke** [comment, marked as answer]
> can change the texture or color or other feature of the objects in scenes. If can, how can we do that ?

Textures cannot currently be changed. For colors, there is a currently undocumented function called `ChangeColorOfMaterials` that could work well. But, it doesn't currently allow you to reset the scenes to their original colors after initially changing randomizing a scene's colors.

Here's a brief example:
<img src="https://user-images.githubusercontent.com/28768645/100716461-96d62300-336d-11eb-9d5f-05bc5bc79189.png" width="50%">

which can be reproduced with
```python
from ai2thor.controller import Controller
import matplotlib.pyplot as plt

c = Controller()
scenes = [f'FloorPlan{i}' for i in ['28', '201', '301']]
fig, axs = plt.subplots(3, 3, figsize=(10, 10))
for i, scene in enumerate(scenes):
    c.reset(scene)
    for j in range(3):
        axs[i,j].imshow(c.last_event.frame)
        axs[i,j].set_xticks([])
        axs[i,j].set_yticks([])

        # change colors
        c.step('ChangeColorOfMaterials')
```

For even crazier effects, it might be worth incorporating [imgaug](https://github.com/aleju/imgaug) or other common augmentations within [torchvision's transforms submodule](https://pytorch.org/docs/stable/torchvision/transforms.html).

> I want to know that if we can add some objects into scenes.

Currently, there is support for duplicating existing objects in the scene (or alternatively removing objects currently in the scene) using `InitialRandomSpawn`, [documented here](https://ai2thor.allenai.org/ithor/documentation/actions/initialization/#object-position-randomization). But, adding in different objects across scenes is not yet supported.

**Reply by liuxz-cs** [reply]
> > can change the texture or color or other feature of the objects in scenes. If can, how can we do that ?
> 
> Textures cannot currently be changed. For colors, there is a currently undocumented function called `ChangeColorOfMaterials` that could work well. But, it doesn't currently allow you to reset the scenes to their original colors after initially changing randomizing a scene's colors.
> 
> Here's a brief example:
> <img alt="" width="50%" src="https://user-images.githubusercontent.com/28768645/100716461-96d62300-336d-11eb-9d5f-05bc5bc79189.png">
> 
> which can be reproduced with
> 
> ```python
> from ai2thor.controller import Controller
> import matplotlib.pyplot as plt
> 
> c = Controller()
> scenes = [f'FloorPlan{i}' for i in ['28', '201', '301']]
> fig, axs = plt.subplots(3, 3, figsize=(10, 10))
> for i, scene in enumerate(scenes):
>     c.reset(scene)
>     for j in range(3):
>         axs[i,j].imshow(c.last_event.frame)
>         axs[i,j].set_xticks([])
>         axs[i,j].set_yticks([])
> 
>         # change colors
>         c.step('ChangeColorOfMaterials')
> ```
> 
> For even crazier effects, it might be worth incorporating [imgaug](https://github.com/aleju/imgaug) or other common augmentations within [torchvision's transforms submodule](https://pytorch.org/docs/stable/torchvision/transforms.html).
> 
> > I want to know that if we can add some objects into scenes.
> 
> Currently, there is support for duplicating existing objects in the scene (or alternatively removing objects currently in the scene) using `InitialRandomSpawn`, [documented here](https://ai2thor.allenai.org/ithor/documentation/actions/initialization/#object-position-randomization). But, adding in different objects across scenes is not yet supported.

So, when we use the function `ChangeColorOfMaterials`, the color is changed randomly ? Can we choose which color to change by this function?

**Reply by mattdeitke** [reply]
> So, when we use the function ChangeColorOfMaterials, the color is changed randomly ?

Correct.

> Can we choose which color to change by this function?

At the moment, no. But, I suppose using instance segmentation masks, one could mask each object individually and then change the hue such that the color changes to some desired color. Although this would be decently expensive from a time perspective.

**Reply by liuxz-cs** [reply]
> > So, when we use the function ChangeColorOfMaterials, the color is changed randomly ?
> 
> Correct.
> 
> > Can we choose which color to change by this function?
> 
> At the moment, no. But, I suppose using instance segmentation masks, one could mask each object individually and then change the hue such that the color changes to some desired color. Although this would be decently expensive from a time perspective.

Okay, thanks a lot.
